documentation/specs/static-graph.md

```diff
@@ -381,3 +457,72 @@ To support this scenario, a new MSBuild Task API could be introduced which allow
+
+```xml
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>netcoreapp3.1</TargetFramework>
+    <UnusualOutput>Configuration\Unusual.txt</UnusualOutput>
+  </PropertyGroup>
+
+  <Target Name="UnusualThing" Returns="$(UnusualOutput)" />
+</Project>
+```
+
+```xml
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>netcoreapp3.1</TargetFramework>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="..\Referenced\Referenced.csproj" />
+  </ItemGroup>
+
+  <Target Name="GetUnusualThing" BeforeTargets="BeforeBuild">
+    <MSBuild Projects="..\Referenced\Referenced.csproj"
+             Targets="UnusualThing">
+      <Output TaskParameter="TargetOutputs"
+              ItemName="Content" />
+    </MSBuild>
+  </Target>
+</Project>
+```
+
+This project can successfully build with `-graph`
+
+```sh-session
+$ dotnet msbuild -graph -noautorsp -nologo
+"Static graph loaded in 0.253 seconds: 2 nodes, 1 edges"
+  You are using a preview version of .NET. See: https://aka.ms/dotnet-core-preview
+  Referenced -> S:\Referenced\bin\Debug\netcoreapp3.1\Referenced.dll
+  Referencing -> S:\Referencing\bin\Debug\netcoreapp3.1\Referencing.dll
+```
+
+But fails with `-graph -isolate`
+
+```sh-session
+$ dotnet msbuild -graph -isolate -noautorsp -nologo
+"Static graph loaded in 0.255 seconds: 2 nodes, 1 edges"
+  You are using a preview version of .NET. See: https://aka.ms/dotnet-core-preview
+  Referenced -> S:\Referenced\bin\Debug\netcoreapp3.1\Referenced.dll
+S:\Referencing\Referencing.csproj(12,5): error : MSB4252: Project "S:\Referencing\Referencing.csproj" with global properties
+S:\Referencing\Referencing.csproj(12,5): error :     (IsGraphBuild=true)
+S:\Referencing\Referencing.csproj(12,5): error :     is building project "S:\Referenced\Referenced.csproj" with global properties
+S:\Referencing\Referencing.csproj(12,5): error :     (IsGraphBuild=true)
+S:\Referencing\Referencing.csproj(12,5): error :     with the (UnusualThing) target(s) but the build result for the built project is not in the engine cache. In isolated builds this could mean one of the following:
+S:\Referencing\Referencing.csproj(12,5): error :     - the reference was called with a target which is not specified in the ProjectReferenceTargets item in project "S:\Referencing\Referencing.csproj"
+S:\Referencing\Referencing.csproj(12,5): error :     - the reference was called with global properties that do not match the static graph inferred nodes
+S:\Referencing\Referencing.csproj(12,5): error :     - the reference was not explicitly specified as a ProjectReference item in project "S:\Referencing\Referencing.csproj"
+S:\Referencing\Referencing.csproj(12,5): error :
+```
+
+This part of the error is the problem here:
+
+> the reference was called with a target which is not specified in the ProjectReferenceTargets item in project "S:\Referencing\Referencing.csproj"
```

Comment:
Think you need to elaborate a bit more on why this is an error in the isolated mode. Particularly because I'm honestly unsure of what the exact problem is here :smile:

Reply:
TLDR: /isolate ensures the graph build does at least the work that vanilla msbuild does. But it's pretty complicated to use and probably only suitable for sdk writers.

- it brutally solves the issue of "what targets should a graph node be called with?". In top-down just-in-time vanilla msbuild a project gets built at the time of discovery. It gets called with whatever arbitrary target values end up in `<MSBuild Targets=$(some_targets)`. During a graph build the runtime graph gets statically inferred and built bottom up, but without doing symbolic execution on the xml we have no idea what targets to call a node with. We solved this by adding a way to statically declare what targets a project calls on its references via the [ProjectReferenceTargets](https://github.com/dotnet/msbuild/blob/master/documentation/specs/static-graph.md#inferring-which-targets-to-run-for-a-project-within-the-graph) protocol, and then we run a data flow over the graph to propagate them. But how do we know the declarations are correct? At minimum the graph build should build at least what vanilla msbuild builds. So one can use /isolate to ensure that the called targets are a superset of the targets run by vanilla msbuild. This is most useful for SDK writers, to ensure that their p2p target calling patterns are inferred by static graph. Not so much useful for end users unless they customize their build.
- correctness wise, even if /isolate is violated, the build would most likely still end up correct. If A calls B with Foo, and Foo is undeclared, then Foo won't get run when the graph build builds B, but will get run regardless when the graph build A, and then A calls into B with Foo. However, if you have a distributed / cached build, it would be nice to capture and cache Foo's potentially expensive CPU work and IO side effects as part of B's scheduled run and cache entry, not as part of A's run. The higher order build engine might even complain if Foo touches the outputs of B, or the outputs of B's references. The fact that Foo's side effects get captured in A's cache entry rather than B's cache entry may or may not be a problem, depending on what one uses the caches for. Just binplacing cached results is fine, doing analyses to infer file dataflow might not be fine.
- MSBuild needs a way to build a project "in isolation", without following references. This is useful when another tool is orchestrating the build. Both VS and Quickbuild build their own msbuild graph and then tell msbuild to build each project in isolation, without following references. To do this both of them rely on the sdks respecting the `BuildProjectReferences=false` convention. /isolate could be used to check that `BuildProjectReferences` is actually respected.
- it enables a potentially interesting way of doing a graph build. If we can correctly infer all the targets a project gets called with, we could visit each node exactly once and call it with all the inferred targets. Then we could serialize those target results and feed them to all the referencing projects. This would enable each project only doing its own work, without reaching out and doing any work in other projects.